// happy-dom/lib/css/CSSStyleSheet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var CSSParser_1 = __importDefault(require("./CSSParser"));
/**
 * Attr node interface.
 */
var CSSStyleSheet = /** @class */ (function () {
    /**
     * Constructor.
     *
     * Constructable Stylesheets is a new feature that only Blink supports:
     * https://wicg.github.io/construct-stylesheets/.
     *
     * @param [options] Options.
     * @param [options.media] Media.
     * @param [options.title] Title.
     * @param [options.alternate] Alternate.
     * @param [options.disabled] Disabled.
     */
    function CSSStyleSheet(options) {
        this.value = null;
        this.name = null;
        this.namespaceURI = null;
        this.cssRules = [];
        this.media = options && options.media ? options.media : '';
        this.title = options && options.title ? options.title : '';
        this.alternate = options && options.alternate ? options.alternate : false;
        this.disabled = options && options.disabled ? options.disabled : false;
    }
    /**
     * Inserts a rule.
     *
     * Constructable Stylesheets is a new feature that only Blink supports:
     * https://wicg.github.io/construct-stylesheets/.
     *
     * @param rule Rule.
     * @param [index] Index.
     */
    CSSStyleSheet.prototype.insertRule = function (rule, index) {
        var _a, _b;
        var rules = CSSParser_1.default.parseFromString(this, rule);
        if (index !== undefined) {
            (_a = this.cssRules).splice.apply(_a, __spreadArray([index, 0], rules));
        }
        else {
            (_b = this.cssRules).push.apply(_b, rules);
        }
    };
    /**
     * Removes a rule.
     *
     * Constructable Stylesheets is a new feature that only Blink supports:
     * https://wicg.github.io/construct-stylesheets/.
     *
     * @param index Index.
     */
    CSSStyleSheet.prototype.deleteRule = function (index) {
        delete this.cssRules[index];
    };
    /**
     * Replaces all CSS rules.
     *
     * Constructable Stylesheets is a new feature that only Blink supports:
     * https://wicg.github.io/construct-stylesheets/.
     *
     * @param text CSS text.
     * @returns Promise.
     */
    CSSStyleSheet.prototype.replace = function (text) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.replaceSync(text);
                return [2 /*return*/];
            });
        });
    };
    /**
     * Replaces all CSS rules.
     *
     * Constructable Stylesheets is a new feature that only Blink supports:
     * https://wicg.github.io/construct-stylesheets/.
     *
     * @param text CSS text.
     */
    CSSStyleSheet.prototype.replaceSync = function (text) {
        this.cssRules = CSSParser_1.default.parseFromString(this, text);
    };
    return CSSStyleSheet;
}());
exports.default = CSSStyleSheet;
