// happy-dom/lib/css/CSSStyleDeclaration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Attr_1 = __importDefault(require("../attribute/Attr"));
/**
 * CSS Style Declaration.
 */
var CSSStyleDeclaration = /** @class */ (function () {
    /**
     * Constructor.
     *
     * @param [attributes] Attributes.
     * @param [computedStyleElement] Computed style element.
     * @param computedStyleElement.isConnected
     */
    function CSSStyleDeclaration(attributes, computedStyleElement) {
        if (attributes === void 0) { attributes = {}; }
        if (computedStyleElement === void 0) { computedStyleElement = null; }
        this.length = 0;
        this.parentRule = null;
        this._attributes = null;
        this._computedStyleElement = null;
        var style = attributes['style'];
        var index = 0;
        this._attributes = attributes;
        this._computedStyleElement = computedStyleElement;
        if (style && style.value) {
            var parts = style.value.split(';');
            for (var _i = 0, parts_1 = parts; _i < parts_1.length; _i++) {
                var part = parts_1[_i];
                if (part) {
                    var name = part.trim().split(':')[0];
                    this[index] = name;
                    index++;
                }
            }
        }
        this.length = index;
    }
    Object.defineProperty(CSSStyleDeclaration.prototype, "alignContent", {
        /* eslint-disable require-jsdoc */
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('align-content');
        },
        /**
         *
         */
        set: function (alignContent) {
            this.setProperty('align-content', alignContent);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "alignItems", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('align-items');
        },
        /**
         *
         */
        set: function (alignItems) {
            this.setProperty('align-items', alignItems);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "alignSelf", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('align-self');
        },
        /**
         *
         */
        set: function (alignSelf) {
            this.setProperty('align-self', alignSelf);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "alignmentBaseline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('alignment-baseline');
        },
        /**
         *
         */
        set: function (alignmentBaseline) {
            this.setProperty('alignment-baseline', alignmentBaseline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "all", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('all');
        },
        /**
         *
         */
        set: function (all) {
            this.setProperty('all', all);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation');
        },
        /**
         *
         */
        set: function (animation) {
            this.setProperty('animation', animation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationDelay", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-delay');
        },
        /**
         *
         */
        set: function (animationDelay) {
            this.setProperty('animation-delay', animationDelay);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationDirection", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-direction');
        },
        /**
         *
         */
        set: function (animationDirection) {
            this.setProperty('animation-direction', animationDirection);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationDuration", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-duration');
        },
        /**
         *
         */
        set: function (animationDuration) {
            this.setProperty('animation-duration', animationDuration);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationFillMode", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-fill-mode');
        },
        /**
         *
         */
        set: function (animationFillMode) {
            this.setProperty('animation-fill-mode', animationFillMode);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationIterationCount", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-iteration-count');
        },
        /**
         *
         */
        set: function (animationIterationCount) {
            this.setProperty('animation-iteration-count', animationIterationCount);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationName", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-name');
        },
        /**
         *
         */
        set: function (animationName) {
            this.setProperty('animation-name', animationName);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationPlayState", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-play-state');
        },
        /**
         *
         */
        set: function (animationPlayState) {
            this.setProperty('animation-play-state', animationPlayState);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "animationTimingFunction", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('animation-timing-function');
        },
        /**
         *
         */
        set: function (animationTimingFunction) {
            this.setProperty('animation-timing-function', animationTimingFunction);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "appearance", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('appearance');
        },
        /**
         *
         */
        set: function (appearance) {
            this.setProperty('appearance', appearance);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backdropFilter", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('backdrop-filter');
        },
        /**
         *
         */
        set: function (backdropFilter) {
            this.setProperty('backdrop-filter', backdropFilter);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backfaceVisibility", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('backface-visibility');
        },
        /**
         *
         */
        set: function (backfaceVisibility) {
            this.setProperty('backface-visibility', backfaceVisibility);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "background", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background');
        },
        /**
         *
         */
        set: function (background) {
            this.setProperty('background', background);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundAttachment", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-attachment');
        },
        /**
         *
         */
        set: function (backgroundAttachment) {
            this.setProperty('background-attachment', backgroundAttachment);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundBlendMode", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-blend-mode');
        },
        /**
         *
         */
        set: function (backgroundBlendMode) {
            this.setProperty('background-blend-mode', backgroundBlendMode);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundClip", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-clip');
        },
        /**
         *
         */
        set: function (backgroundClip) {
            this.setProperty('background-clip', backgroundClip);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-color');
        },
        /**
         *
         */
        set: function (backgroundColor) {
            this.setProperty('background-color', backgroundColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundImage", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-image');
        },
        /**
         *
         */
        set: function (backgroundImage) {
            this.setProperty('background-image', backgroundImage);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundOrigin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-origin');
        },
        /**
         *
         */
        set: function (backgroundOrigin) {
            this.setProperty('background-origin', backgroundOrigin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundPosition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-position');
        },
        /**
         *
         */
        set: function (backgroundPosition) {
            this.setProperty('background-position', backgroundPosition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundPositionX", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-position-x');
        },
        /**
         *
         */
        set: function (backgroundPositionX) {
            this.setProperty('background-position-x', backgroundPositionX);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundPositionY", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-position-y');
        },
        /**
         *
         */
        set: function (backgroundPositionY) {
            this.setProperty('background-position-y', backgroundPositionY);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundRepeat", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-repeat');
        },
        /**
         *
         */
        set: function (backgroundRepeat) {
            this.setProperty('background-repeat', backgroundRepeat);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundRepeatX", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-repeat-x');
        },
        /**
         *
         */
        set: function (backgroundRepeatX) {
            this.setProperty('background-repeat-x', backgroundRepeatX);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundRepeatY", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-repeat-y');
        },
        /**
         *
         */
        set: function (backgroundRepeatY) {
            this.setProperty('background-repeat-y', backgroundRepeatY);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "backgroundSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('background-size');
        },
        /**
         *
         */
        set: function (backgroundSize) {
            this.setProperty('background-size', backgroundSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "baselineShift", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('baseline-shift');
        },
        /**
         *
         */
        set: function (baselineShift) {
            this.setProperty('baseline-shift', baselineShift);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "blockSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('block-size');
        },
        /**
         *
         */
        set: function (blockSize) {
            this.setProperty('block-size', blockSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "border", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border');
        },
        /**
         *
         */
        set: function (border) {
            this.setProperty('border', border);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-end');
        },
        /**
         *
         */
        set: function (borderBlockEnd) {
            this.setProperty('border-block-end', borderBlockEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockEndColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-end-color');
        },
        /**
         *
         */
        set: function (borderBlockEndColor) {
            this.setProperty('border-block-end-color', borderBlockEndColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockEndStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-end-style');
        },
        /**
         *
         */
        set: function (borderBlockEndStyle) {
            this.setProperty('border-block-end-style', borderBlockEndStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockEndWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-end-width');
        },
        /**
         *
         */
        set: function (borderBlockEndWidth) {
            this.setProperty('border-block-end-width', borderBlockEndWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-start');
        },
        /**
         *
         */
        set: function (borderBlockStart) {
            this.setProperty('border-block-start', borderBlockStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockStartColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-start-color');
        },
        /**
         *
         */
        set: function (borderBlockStartColor) {
            this.setProperty('border-block-start-color', borderBlockStartColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockStartStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-start-style');
        },
        /**
         *
         */
        set: function (borderBlockStartStyle) {
            this.setProperty('border-block-start-style', borderBlockStartStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBlockStartWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-block-start-width');
        },
        /**
         *
         */
        set: function (borderBlockStartWidth) {
            this.setProperty('border-block-start-width', borderBlockStartWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom');
        },
        /**
         *
         */
        set: function (borderBottom) {
            this.setProperty('border-bottom', borderBottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottomColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom-color');
        },
        /**
         *
         */
        set: function (borderBottomColor) {
            this.setProperty('border-bottom-color', borderBottomColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottomLeftRadius", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom-left-radius');
        },
        /**
         *
         */
        set: function (borderBottomLeftRadius) {
            this.setProperty('border-bottom-left-radius', borderBottomLeftRadius);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottomRightRadius", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom-right-radius');
        },
        /**
         *
         */
        set: function (borderBottomRightRadius) {
            this.setProperty('border-bottom-right-radius', borderBottomRightRadius);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottomStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom-style');
        },
        /**
         *
         */
        set: function (borderBottomStyle) {
            this.setProperty('border-bottom-style', borderBottomStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderBottomWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-bottom-width');
        },
        /**
         *
         */
        set: function (borderBottomWidth) {
            this.setProperty('border-bottom-width', borderBottomWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderCollapse", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-collapse');
        },
        /**
         *
         */
        set: function (borderCollapse) {
            this.setProperty('border-collapse', borderCollapse);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-color');
        },
        /**
         *
         */
        set: function (borderColor) {
            this.setProperty('border-color', borderColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImage", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image');
        },
        /**
         *
         */
        set: function (borderImage) {
            this.setProperty('border-image', borderImage);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImageOutset", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image-outset');
        },
        /**
         *
         */
        set: function (borderImageOutset) {
            this.setProperty('border-image-outset', borderImageOutset);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImageRepeat", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image-repeat');
        },
        /**
         *
         */
        set: function (borderImageRepeat) {
            this.setProperty('border-image-repeat', borderImageRepeat);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImageSlice", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image-slice');
        },
        /**
         *
         */
        set: function (borderImageSlice) {
            this.setProperty('border-image-slice', borderImageSlice);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImageSource", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image-source');
        },
        /**
         *
         */
        set: function (borderImageSource) {
            this.setProperty('border-image-source', borderImageSource);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderImageWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-image-width');
        },
        /**
         *
         */
        set: function (borderImageWidth) {
            this.setProperty('border-image-width', borderImageWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-end');
        },
        /**
         *
         */
        set: function (borderInlineEnd) {
            this.setProperty('border-inline-end', borderInlineEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineEndColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-end-color');
        },
        /**
         *
         */
        set: function (borderInlineEndColor) {
            this.setProperty('border-inline-end-color', borderInlineEndColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineEndStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-end-style');
        },
        /**
         *
         */
        set: function (borderInlineEndStyle) {
            this.setProperty('border-inline-end-style', borderInlineEndStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineEndWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-end-width');
        },
        /**
         *
         */
        set: function (borderInlineEndWidth) {
            this.setProperty('border-inline-end-width', borderInlineEndWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-start');
        },
        /**
         *
         */
        set: function (borderInlineStart) {
            this.setProperty('border-inline-start', borderInlineStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineStartColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-start-color');
        },
        /**
         *
         */
        set: function (borderInlineStartColor) {
            this.setProperty('border-inline-start-color', borderInlineStartColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineStartStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-start-style');
        },
        /**
         *
         */
        set: function (borderInlineStartStyle) {
            this.setProperty('border-inline-start-style', borderInlineStartStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderInlineStartWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-inline-start-width');
        },
        /**
         *
         */
        set: function (borderInlineStartWidth) {
            this.setProperty('border-inline-start-width', borderInlineStartWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderLeft", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-left');
        },
        /**
         *
         */
        set: function (borderLeft) {
            this.setProperty('border-left', borderLeft);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderLeftColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-left-color');
        },
        /**
         *
         */
        set: function (borderLeftColor) {
            this.setProperty('border-left-color', borderLeftColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderLeftStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-left-style');
        },
        /**
         *
         */
        set: function (borderLeftStyle) {
            this.setProperty('border-left-style', borderLeftStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderLeftWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-left-width');
        },
        /**
         *
         */
        set: function (borderLeftWidth) {
            this.setProperty('border-left-width', borderLeftWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderRadius", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-radius');
        },
        /**
         *
         */
        set: function (borderRadius) {
            this.setProperty('border-radius', borderRadius);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderRight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-right');
        },
        /**
         *
         */
        set: function (borderRight) {
            this.setProperty('border-right', borderRight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderRightColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-right-color');
        },
        /**
         *
         */
        set: function (borderRightColor) {
            this.setProperty('border-right-color', borderRightColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderRightStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-right-style');
        },
        /**
         *
         */
        set: function (borderRightStyle) {
            this.setProperty('border-right-style', borderRightStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderRightWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-right-width');
        },
        /**
         *
         */
        set: function (borderRightWidth) {
            this.setProperty('border-right-width', borderRightWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderSpacing", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-spacing');
        },
        /**
         *
         */
        set: function (borderSpacing) {
            this.setProperty('border-spacing', borderSpacing);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-style');
        },
        /**
         *
         */
        set: function (borderStyle) {
            this.setProperty('border-style', borderStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top');
        },
        /**
         *
         */
        set: function (borderTop) {
            this.setProperty('border-top', borderTop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTopColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top-color');
        },
        /**
         *
         */
        set: function (borderTopColor) {
            this.setProperty('border-top-color', borderTopColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTopLeftRadius", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top-left-radius');
        },
        /**
         *
         */
        set: function (borderTopLeftRadius) {
            this.setProperty('border-top-left-radius', borderTopLeftRadius);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTopRightRadius", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top-right-radius');
        },
        /**
         *
         */
        set: function (borderTopRightRadius) {
            this.setProperty('border-top-right-radius', borderTopRightRadius);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTopStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top-style');
        },
        /**
         *
         */
        set: function (borderTopStyle) {
            this.setProperty('border-top-style', borderTopStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderTopWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-top-width');
        },
        /**
         *
         */
        set: function (borderTopWidth) {
            this.setProperty('border-top-width', borderTopWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "borderWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('border-width');
        },
        /**
         *
         */
        set: function (borderWidth) {
            this.setProperty('border-width', borderWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "bottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('bottom');
        },
        /**
         *
         */
        set: function (bottom) {
            this.setProperty('bottom', bottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "boxShadow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('box-shadow');
        },
        /**
         *
         */
        set: function (boxShadow) {
            this.setProperty('box-shadow', boxShadow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "boxSizing", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('box-sizing');
        },
        /**
         *
         */
        set: function (boxSizing) {
            this.setProperty('box-sizing', boxSizing);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "breakAfter", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('break-after');
        },
        /**
         *
         */
        set: function (breakAfter) {
            this.setProperty('break-after', breakAfter);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "breakBefore", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('break-before');
        },
        /**
         *
         */
        set: function (breakBefore) {
            this.setProperty('break-before', breakBefore);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "breakInside", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('break-inside');
        },
        /**
         *
         */
        set: function (breakInside) {
            this.setProperty('break-inside', breakInside);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "bufferedRendering", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('buffered-rendering');
        },
        /**
         *
         */
        set: function (bufferedRendering) {
            this.setProperty('buffered-rendering', bufferedRendering);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "captionSide", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('caption-side');
        },
        /**
         *
         */
        set: function (captionSide) {
            this.setProperty('caption-side', captionSide);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "caretColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('caret-color');
        },
        /**
         *
         */
        set: function (caretColor) {
            this.setProperty('caret-color', caretColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "clear", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('clear');
        },
        /**
         *
         */
        set: function (clear) {
            this.setProperty('clear', clear);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "clip", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('clip');
        },
        /**
         *
         */
        set: function (clip) {
            this.setProperty('clip', clip);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "clipPath", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('clip-path');
        },
        /**
         *
         */
        set: function (clipPath) {
            this.setProperty('clip-path', clipPath);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "clipRule", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('clip-rule');
        },
        /**
         *
         */
        set: function (clipRule) {
            this.setProperty('clip-rule', clipRule);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "color", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('color');
        },
        /**
         *
         */
        set: function (color) {
            this.setProperty('color', color);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "colorInterpolation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('color-interpolation');
        },
        /**
         *
         */
        set: function (colorInterpolation) {
            this.setProperty('color-interpolation', colorInterpolation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "colorInterpolationFilters", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('color-interpolation-filters');
        },
        /**
         *
         */
        set: function (colorInterpolationFilters) {
            this.setProperty('color-interpolation-filters', colorInterpolationFilters);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "colorRendering", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('color-rendering');
        },
        /**
         *
         */
        set: function (colorRendering) {
            this.setProperty('color-rendering', colorRendering);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "colorScheme", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('color-scheme');
        },
        /**
         *
         */
        set: function (colorScheme) {
            this.setProperty('color-scheme', colorScheme);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnCount", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-count');
        },
        /**
         *
         */
        set: function (columnCount) {
            this.setProperty('column-count', columnCount);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnFill", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-fill');
        },
        /**
         *
         */
        set: function (columnFill) {
            this.setProperty('column-fill', columnFill);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnGap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-gap');
        },
        /**
         *
         */
        set: function (columnGap) {
            this.setProperty('column-gap', columnGap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnRule", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-rule');
        },
        /**
         *
         */
        set: function (columnRule) {
            this.setProperty('column-rule', columnRule);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnRuleColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-rule-color');
        },
        /**
         *
         */
        set: function (columnRuleColor) {
            this.setProperty('column-rule-color', columnRuleColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnRuleStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-rule-style');
        },
        /**
         *
         */
        set: function (columnRuleStyle) {
            this.setProperty('column-rule-style', columnRuleStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnRuleWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-rule-width');
        },
        /**
         *
         */
        set: function (columnRuleWidth) {
            this.setProperty('column-rule-width', columnRuleWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnSpan", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-span');
        },
        /**
         *
         */
        set: function (columnSpan) {
            this.setProperty('column-span', columnSpan);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columnWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('column-width');
        },
        /**
         *
         */
        set: function (columnWidth) {
            this.setProperty('column-width', columnWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "columns", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('columns');
        },
        /**
         *
         */
        set: function (columns) {
            this.setProperty('columns', columns);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "contain", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('contain');
        },
        /**
         *
         */
        set: function (contain) {
            this.setProperty('contain', contain);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "containIntrinsicSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('contain-intrinsic-size');
        },
        /**
         *
         */
        set: function (containIntrinsicSize) {
            this.setProperty('contain-intrinsic-size', containIntrinsicSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "content", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('content');
        },
        /**
         *
         */
        set: function (content) {
            this.setProperty('content', content);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "contentVisibility", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('content-visibility');
        },
        /**
         *
         */
        set: function (contentVisibility) {
            this.setProperty('content-visibility', contentVisibility);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "counterIncrement", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('counter-increment');
        },
        /**
         *
         */
        set: function (counterIncrement) {
            this.setProperty('counter-increment', counterIncrement);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "counterReset", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('counter-reset');
        },
        /**
         *
         */
        set: function (counterReset) {
            this.setProperty('counter-reset', counterReset);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "counterSet", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('counter-set');
        },
        /**
         *
         */
        set: function (counterSet) {
            this.setProperty('counter-set', counterSet);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "cssFloat", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('css-float');
        },
        /**
         *
         */
        set: function (cssFloat) {
            this.setProperty('css-float', cssFloat);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "cursor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('cursor');
        },
        /**
         *
         */
        set: function (cursor) {
            this.setProperty('cursor', cursor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "cx", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('cx');
        },
        /**
         *
         */
        set: function (cx) {
            this.setProperty('cx', cx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "cy", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('cy');
        },
        /**
         *
         */
        set: function (cy) {
            this.setProperty('cy', cy);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "d", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('d');
        },
        /**
         *
         */
        set: function (d) {
            this.setProperty('d', d);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "direction", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('direction');
        },
        /**
         *
         */
        set: function (direction) {
            this.setProperty('direction', direction);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "display", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('display');
        },
        /**
         *
         */
        set: function (display) {
            this.setProperty('display', display);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "dominantBaseline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('dominant-baseline');
        },
        /**
         *
         */
        set: function (dominantBaseline) {
            this.setProperty('dominant-baseline', dominantBaseline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "emptyCells", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('empty-cells');
        },
        /**
         *
         */
        set: function (emptyCells) {
            this.setProperty('empty-cells', emptyCells);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fill", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('fill');
        },
        /**
         *
         */
        set: function (fill) {
            this.setProperty('fill', fill);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fillOpacity", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('fill-opacity');
        },
        /**
         *
         */
        set: function (fillOpacity) {
            this.setProperty('fill-opacity', fillOpacity);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fillRule", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('fill-rule');
        },
        /**
         *
         */
        set: function (fillRule) {
            this.setProperty('fill-rule', fillRule);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "filter", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('filter');
        },
        /**
         *
         */
        set: function (filter) {
            this.setProperty('filter', filter);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flex", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex');
        },
        /**
         *
         */
        set: function (flex) {
            this.setProperty('flex', flex);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexBasis", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-basis');
        },
        /**
         *
         */
        set: function (flexBasis) {
            this.setProperty('flex-basis', flexBasis);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexDirection", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-direction');
        },
        /**
         *
         */
        set: function (flexDirection) {
            this.setProperty('flex-direction', flexDirection);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexFlow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-flow');
        },
        /**
         *
         */
        set: function (flexFlow) {
            this.setProperty('flex-flow', flexFlow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexGrow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-grow');
        },
        /**
         *
         */
        set: function (flexGrow) {
            this.setProperty('flex-grow', flexGrow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexShrink", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-shrink');
        },
        /**
         *
         */
        set: function (flexShrink) {
            this.setProperty('flex-shrink', flexShrink);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "flexWrap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flex-wrap');
        },
        /**
         *
         */
        set: function (flexWrap) {
            this.setProperty('flex-wrap', flexWrap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "float", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('float');
        },
        /**
         *
         */
        set: function (float) {
            this.setProperty('float', float);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "floodColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flood-color');
        },
        /**
         *
         */
        set: function (floodColor) {
            this.setProperty('flood-color', floodColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "floodOpacity", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('flood-opacity');
        },
        /**
         *
         */
        set: function (floodOpacity) {
            this.setProperty('flood-opacity', floodOpacity);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "font", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font');
        },
        /**
         *
         */
        set: function (font) {
            this.setProperty('font', font);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontDisplay", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-display');
        },
        /**
         *
         */
        set: function (fontDisplay) {
            this.setProperty('font-display', fontDisplay);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontFamily", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-family');
        },
        /**
         *
         */
        set: function (fontFamily) {
            this.setProperty('font-family', fontFamily);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontFeatureSettings", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-feature-settings');
        },
        /**
         *
         */
        set: function (fontFeatureSettings) {
            this.setProperty('font-feature-settings', fontFeatureSettings);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontKerning", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-kerning');
        },
        /**
         *
         */
        set: function (fontKerning) {
            this.setProperty('font-kerning', fontKerning);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontOpticalSizing", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-optical-sizing');
        },
        /**
         *
         */
        set: function (fontOpticalSizing) {
            this.setProperty('font-optical-sizing', fontOpticalSizing);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-size');
        },
        /**
         *
         */
        set: function (fontSize) {
            this.setProperty('font-size', fontSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontStretch", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-stretch');
        },
        /**
         *
         */
        set: function (fontStretch) {
            this.setProperty('font-stretch', fontStretch);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-style');
        },
        /**
         *
         */
        set: function (fontStyle) {
            this.setProperty('font-style', fontStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariant", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variant');
        },
        /**
         *
         */
        set: function (fontVariant) {
            this.setProperty('font-variant', fontVariant);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariantCaps", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variant-caps');
        },
        /**
         *
         */
        set: function (fontVariantCaps) {
            this.setProperty('font-variant-caps', fontVariantCaps);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariantEastAsian", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variant-east-asian');
        },
        /**
         *
         */
        set: function (fontVariantEastAsian) {
            this.setProperty('font-variant-east-asian', fontVariantEastAsian);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariantLigatures", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variant-ligatures');
        },
        /**
         *
         */
        set: function (fontVariantLigatures) {
            this.setProperty('font-variant-ligatures', fontVariantLigatures);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariantNumeric", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variant-numeric');
        },
        /**
         *
         */
        set: function (fontVariantNumeric) {
            this.setProperty('font-variant-numeric', fontVariantNumeric);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontVariationSettings", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-variation-settings');
        },
        /**
         *
         */
        set: function (fontVariationSettings) {
            this.setProperty('font-variation-settings', fontVariationSettings);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "fontWeight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('font-weight');
        },
        /**
         *
         */
        set: function (fontWeight) {
            this.setProperty('font-weight', fontWeight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('gap');
        },
        /**
         *
         */
        set: function (gap) {
            this.setProperty('gap', gap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "grid", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid');
        },
        /**
         *
         */
        set: function (grid) {
            this.setProperty('grid', grid);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridArea", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-area');
        },
        /**
         *
         */
        set: function (gridArea) {
            this.setProperty('grid-area', gridArea);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridAutoColumns", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-auto-columns');
        },
        /**
         *
         */
        set: function (gridAutoColumns) {
            this.setProperty('grid-auto-columns', gridAutoColumns);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridAutoFlow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-auto-flow');
        },
        /**
         *
         */
        set: function (gridAutoFlow) {
            this.setProperty('grid-auto-flow', gridAutoFlow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridAutoRows", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-auto-rows');
        },
        /**
         *
         */
        set: function (gridAutoRows) {
            this.setProperty('grid-auto-rows', gridAutoRows);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridColumn", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-column');
        },
        /**
         *
         */
        set: function (gridColumn) {
            this.setProperty('grid-column', gridColumn);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridColumnEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-column-end');
        },
        /**
         *
         */
        set: function (gridColumnEnd) {
            this.setProperty('grid-column-end', gridColumnEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridColumnGap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-column-gap');
        },
        /**
         *
         */
        set: function (gridColumnGap) {
            this.setProperty('grid-column-gap', gridColumnGap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridColumnStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-column-start');
        },
        /**
         *
         */
        set: function (gridColumnStart) {
            this.setProperty('grid-column-start', gridColumnStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridGap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-gap');
        },
        /**
         *
         */
        set: function (gridGap) {
            this.setProperty('grid-gap', gridGap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridRow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-row');
        },
        /**
         *
         */
        set: function (gridRow) {
            this.setProperty('grid-row', gridRow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridRowEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-row-end');
        },
        /**
         *
         */
        set: function (gridRowEnd) {
            this.setProperty('grid-row-end', gridRowEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridRowGap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-row-gap');
        },
        /**
         *
         */
        set: function (gridRowGap) {
            this.setProperty('grid-row-gap', gridRowGap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridRowStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-row-start');
        },
        /**
         *
         */
        set: function (gridRowStart) {
            this.setProperty('grid-row-start', gridRowStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridTemplate", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-template');
        },
        /**
         *
         */
        set: function (gridTemplate) {
            this.setProperty('grid-template', gridTemplate);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridTemplateAreas", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-template-areas');
        },
        /**
         *
         */
        set: function (gridTemplateAreas) {
            this.setProperty('grid-template-areas', gridTemplateAreas);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridTemplateColumns", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-template-columns');
        },
        /**
         *
         */
        set: function (gridTemplateColumns) {
            this.setProperty('grid-template-columns', gridTemplateColumns);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "gridTemplateRows", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('grid-template-rows');
        },
        /**
         *
         */
        set: function (gridTemplateRows) {
            this.setProperty('grid-template-rows', gridTemplateRows);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "height", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('height');
        },
        /**
         *
         */
        set: function (height) {
            this.setProperty('height', height);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "hyphens", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('hyphens');
        },
        /**
         *
         */
        set: function (hyphens) {
            this.setProperty('hyphens', hyphens);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "imageOrientation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('image-orientation');
        },
        /**
         *
         */
        set: function (imageOrientation) {
            this.setProperty('image-orientation', imageOrientation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "imageRendering", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('image-rendering');
        },
        /**
         *
         */
        set: function (imageRendering) {
            this.setProperty('image-rendering', imageRendering);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "inherits", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('inherits');
        },
        /**
         *
         */
        set: function (inherits) {
            this.setProperty('inherits', inherits);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "initialValue", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('initial-value');
        },
        /**
         *
         */
        set: function (initialValue) {
            this.setProperty('initial-value', initialValue);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "inlineSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('inline-size');
        },
        /**
         *
         */
        set: function (inlineSize) {
            this.setProperty('inline-size', inlineSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "isolation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('isolation');
        },
        /**
         *
         */
        set: function (isolation) {
            this.setProperty('isolation', isolation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "justifyContent", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('justify-content');
        },
        /**
         *
         */
        set: function (justifyContent) {
            this.setProperty('justify-content', justifyContent);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "justifyItems", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('justify-items');
        },
        /**
         *
         */
        set: function (justifyItems) {
            this.setProperty('justify-items', justifyItems);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "justifySelf", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('justify-self');
        },
        /**
         *
         */
        set: function (justifySelf) {
            this.setProperty('justify-self', justifySelf);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "left", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('left');
        },
        /**
         *
         */
        set: function (left) {
            this.setProperty('left', left);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "letterSpacing", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('letter-spacing');
        },
        /**
         *
         */
        set: function (letterSpacing) {
            this.setProperty('letter-spacing', letterSpacing);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "lightingColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('lighting-color');
        },
        /**
         *
         */
        set: function (lightingColor) {
            this.setProperty('lighting-color', lightingColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "lineBreak", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('line-break');
        },
        /**
         *
         */
        set: function (lineBreak) {
            this.setProperty('line-break', lineBreak);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "lineHeight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('line-height');
        },
        /**
         *
         */
        set: function (lineHeight) {
            this.setProperty('line-height', lineHeight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "listStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('list-style');
        },
        /**
         *
         */
        set: function (listStyle) {
            this.setProperty('list-style', listStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "listStyleImage", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('list-style-image');
        },
        /**
         *
         */
        set: function (listStyleImage) {
            this.setProperty('list-style-image', listStyleImage);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "listStylePosition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('list-style-position');
        },
        /**
         *
         */
        set: function (listStylePosition) {
            this.setProperty('list-style-position', listStylePosition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "listStyleType", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('list-style-type');
        },
        /**
         *
         */
        set: function (listStyleType) {
            this.setProperty('list-style-type', listStyleType);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "margin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin');
        },
        /**
         *
         */
        set: function (margin) {
            this.setProperty('margin', margin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginBlockEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-block-end');
        },
        /**
         *
         */
        set: function (marginBlockEnd) {
            this.setProperty('margin-block-end', marginBlockEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginBlockStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-block-start');
        },
        /**
         *
         */
        set: function (marginBlockStart) {
            this.setProperty('margin-block-start', marginBlockStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginBottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-bottom');
        },
        /**
         *
         */
        set: function (marginBottom) {
            this.setProperty('margin-bottom', marginBottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginInlineEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-inline-end');
        },
        /**
         *
         */
        set: function (marginInlineEnd) {
            this.setProperty('margin-inline-end', marginInlineEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginInlineStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-inline-start');
        },
        /**
         *
         */
        set: function (marginInlineStart) {
            this.setProperty('margin-inline-start', marginInlineStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginLeft", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-left');
        },
        /**
         *
         */
        set: function (marginLeft) {
            this.setProperty('margin-left', marginLeft);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginRight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-right');
        },
        /**
         *
         */
        set: function (marginRight) {
            this.setProperty('margin-right', marginRight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marginTop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('margin-top');
        },
        /**
         *
         */
        set: function (marginTop) {
            this.setProperty('margin-top', marginTop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "marker", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('marker');
        },
        /**
         *
         */
        set: function (marker) {
            this.setProperty('marker', marker);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "markerEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('marker-end');
        },
        /**
         *
         */
        set: function (markerEnd) {
            this.setProperty('marker-end', markerEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "markerMid", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('marker-mid');
        },
        /**
         *
         */
        set: function (markerMid) {
            this.setProperty('marker-mid', markerMid);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "markerStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('marker-start');
        },
        /**
         *
         */
        set: function (markerStart) {
            this.setProperty('marker-start', markerStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "mask", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('mask');
        },
        /**
         *
         */
        set: function (mask) {
            this.setProperty('mask', mask);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maskType", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('mask-type');
        },
        /**
         *
         */
        set: function (maskType) {
            this.setProperty('mask-type', maskType);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maxBlockSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('max-block-size');
        },
        /**
         *
         */
        set: function (maxBlockSize) {
            this.setProperty('max-block-size', maxBlockSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maxHeight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('max-height');
        },
        /**
         *
         */
        set: function (maxHeight) {
            this.setProperty('max-height', maxHeight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maxInlineSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('max-inline-size');
        },
        /**
         *
         */
        set: function (maxInlineSize) {
            this.setProperty('max-inline-size', maxInlineSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maxWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('max-width');
        },
        /**
         *
         */
        set: function (maxWidth) {
            this.setProperty('max-width', maxWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "maxZoom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('max-zoom');
        },
        /**
         *
         */
        set: function (maxZoom) {
            this.setProperty('max-zoom', maxZoom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "minBlockSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('min-block-size');
        },
        /**
         *
         */
        set: function (minBlockSize) {
            this.setProperty('min-block-size', minBlockSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "minHeight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('min-height');
        },
        /**
         *
         */
        set: function (minHeight) {
            this.setProperty('min-height', minHeight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "minInlineSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('min-inline-size');
        },
        /**
         *
         */
        set: function (minInlineSize) {
            this.setProperty('min-inline-size', minInlineSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "minWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('min-width');
        },
        /**
         *
         */
        set: function (minWidth) {
            this.setProperty('min-width', minWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "minZoom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('min-zoom');
        },
        /**
         *
         */
        set: function (minZoom) {
            this.setProperty('min-zoom', minZoom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "mixBlendMode", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('mix-blend-mode');
        },
        /**
         *
         */
        set: function (mixBlendMode) {
            this.setProperty('mix-blend-mode', mixBlendMode);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "objectFit", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('object-fit');
        },
        /**
         *
         */
        set: function (objectFit) {
            this.setProperty('object-fit', objectFit);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "objectPosition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('object-position');
        },
        /**
         *
         */
        set: function (objectPosition) {
            this.setProperty('object-position', objectPosition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "offset", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('offset');
        },
        /**
         *
         */
        set: function (offset) {
            this.setProperty('offset', offset);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "offsetDistance", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('offset-distance');
        },
        /**
         *
         */
        set: function (offsetDistance) {
            this.setProperty('offset-distance', offsetDistance);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "offsetPath", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('offset-path');
        },
        /**
         *
         */
        set: function (offsetPath) {
            this.setProperty('offset-path', offsetPath);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "offsetRotate", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('offset-rotate');
        },
        /**
         *
         */
        set: function (offsetRotate) {
            this.setProperty('offset-rotate', offsetRotate);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "opacity", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('opacity');
        },
        /**
         *
         */
        set: function (opacity) {
            this.setProperty('opacity', opacity);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "order", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('order');
        },
        /**
         *
         */
        set: function (order) {
            this.setProperty('order', order);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "orientation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('orientation');
        },
        /**
         *
         */
        set: function (orientation) {
            this.setProperty('orientation', orientation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "orphans", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('orphans');
        },
        /**
         *
         */
        set: function (orphans) {
            this.setProperty('orphans', orphans);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "outline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('outline');
        },
        /**
         *
         */
        set: function (outline) {
            this.setProperty('outline', outline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "outlineColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('outline-color');
        },
        /**
         *
         */
        set: function (outlineColor) {
            this.setProperty('outline-color', outlineColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "outlineOffset", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('outline-offset');
        },
        /**
         *
         */
        set: function (outlineOffset) {
            this.setProperty('outline-offset', outlineOffset);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "outlineStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('outline-style');
        },
        /**
         *
         */
        set: function (outlineStyle) {
            this.setProperty('outline-style', outlineStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "outlineWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('outline-width');
        },
        /**
         *
         */
        set: function (outlineWidth) {
            this.setProperty('outline-width', outlineWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overflow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overflow');
        },
        /**
         *
         */
        set: function (overflow) {
            this.setProperty('overflow', overflow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overflowAnchor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overflow-anchor');
        },
        /**
         *
         */
        set: function (overflowAnchor) {
            this.setProperty('overflow-anchor', overflowAnchor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overflowWrap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overflow-wrap');
        },
        /**
         *
         */
        set: function (overflowWrap) {
            this.setProperty('overflow-wrap', overflowWrap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overflowX", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overflow-x');
        },
        /**
         *
         */
        set: function (overflowX) {
            this.setProperty('overflow-x', overflowX);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overflowY", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overflow-y');
        },
        /**
         *
         */
        set: function (overflowY) {
            this.setProperty('overflow-y', overflowY);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overscrollBehavior", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overscroll-behavior');
        },
        /**
         *
         */
        set: function (overscrollBehavior) {
            this.setProperty('overscroll-behavior', overscrollBehavior);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overscrollBehaviorBlock", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overscroll-behavior-block');
        },
        /**
         *
         */
        set: function (overscrollBehaviorBlock) {
            this.setProperty('overscroll-behavior-block', overscrollBehaviorBlock);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overscrollBehaviorInline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overscroll-behavior-inline');
        },
        /**
         *
         */
        set: function (overscrollBehaviorInline) {
            this.setProperty('overscroll-behavior-inline', overscrollBehaviorInline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overscrollBehaviorX", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overscroll-behavior-x');
        },
        /**
         *
         */
        set: function (overscrollBehaviorX) {
            this.setProperty('overscroll-behavior-x', overscrollBehaviorX);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "overscrollBehaviorY", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('overscroll-behavior-y');
        },
        /**
         *
         */
        set: function (overscrollBehaviorY) {
            this.setProperty('overscroll-behavior-y', overscrollBehaviorY);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "padding", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding');
        },
        /**
         *
         */
        set: function (padding) {
            this.setProperty('padding', padding);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingBlockEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-block-end');
        },
        /**
         *
         */
        set: function (paddingBlockEnd) {
            this.setProperty('padding-block-end', paddingBlockEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingBlockStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-block-start');
        },
        /**
         *
         */
        set: function (paddingBlockStart) {
            this.setProperty('padding-block-start', paddingBlockStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingBottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-bottom');
        },
        /**
         *
         */
        set: function (paddingBottom) {
            this.setProperty('padding-bottom', paddingBottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingInlineEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-inline-end');
        },
        /**
         *
         */
        set: function (paddingInlineEnd) {
            this.setProperty('padding-inline-end', paddingInlineEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingInlineStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-inline-start');
        },
        /**
         *
         */
        set: function (paddingInlineStart) {
            this.setProperty('padding-inline-start', paddingInlineStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingLeft", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-left');
        },
        /**
         *
         */
        set: function (paddingLeft) {
            this.setProperty('padding-left', paddingLeft);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingRight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-right');
        },
        /**
         *
         */
        set: function (paddingRight) {
            this.setProperty('padding-right', paddingRight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paddingTop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('padding-top');
        },
        /**
         *
         */
        set: function (paddingTop) {
            this.setProperty('padding-top', paddingTop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "page", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('page');
        },
        /**
         *
         */
        set: function (page) {
            this.setProperty('page', page);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "pageBreakAfter", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('page-break-after');
        },
        /**
         *
         */
        set: function (pageBreakAfter) {
            this.setProperty('page-break-after', pageBreakAfter);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "pageBreakBefore", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('page-break-before');
        },
        /**
         *
         */
        set: function (pageBreakBefore) {
            this.setProperty('page-break-before', pageBreakBefore);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "pageBreakInside", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('page-break-inside');
        },
        /**
         *
         */
        set: function (pageBreakInside) {
            this.setProperty('page-break-inside', pageBreakInside);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "pageOrientation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('page-orientation');
        },
        /**
         *
         */
        set: function (pageOrientation) {
            this.setProperty('page-orientation', pageOrientation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "paintOrder", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('paint-order');
        },
        /**
         *
         */
        set: function (paintOrder) {
            this.setProperty('paint-order', paintOrder);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "perspective", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('perspective');
        },
        /**
         *
         */
        set: function (perspective) {
            this.setProperty('perspective', perspective);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "perspectiveOrigin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('perspective-origin');
        },
        /**
         *
         */
        set: function (perspectiveOrigin) {
            this.setProperty('perspective-origin', perspectiveOrigin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "placeContent", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('place-content');
        },
        /**
         *
         */
        set: function (placeContent) {
            this.setProperty('place-content', placeContent);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "placeItems", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('place-items');
        },
        /**
         *
         */
        set: function (placeItems) {
            this.setProperty('place-items', placeItems);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "placeSelf", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('place-self');
        },
        /**
         *
         */
        set: function (placeSelf) {
            this.setProperty('place-self', placeSelf);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "pointerEvents", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('pointer-events');
        },
        /**
         *
         */
        set: function (pointerEvents) {
            this.setProperty('pointer-events', pointerEvents);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "position", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('position');
        },
        /**
         *
         */
        set: function (position) {
            this.setProperty('position', position);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "quotes", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('quotes');
        },
        /**
         *
         */
        set: function (quotes) {
            this.setProperty('quotes', quotes);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "r", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('r');
        },
        /**
         *
         */
        set: function (r) {
            this.setProperty('r', r);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "resize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('resize');
        },
        /**
         *
         */
        set: function (resize) {
            this.setProperty('resize', resize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "right", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('right');
        },
        /**
         *
         */
        set: function (right) {
            this.setProperty('right', right);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "rowGap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('row-gap');
        },
        /**
         *
         */
        set: function (rowGap) {
            this.setProperty('row-gap', rowGap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "rubyPosition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('ruby-position');
        },
        /**
         *
         */
        set: function (rubyPosition) {
            this.setProperty('ruby-position', rubyPosition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "rx", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('rx');
        },
        /**
         *
         */
        set: function (rx) {
            this.setProperty('rx', rx);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "ry", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('ry');
        },
        /**
         *
         */
        set: function (ry) {
            this.setProperty('ry', ry);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollBehavior", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-behavior');
        },
        /**
         *
         */
        set: function (scrollBehavior) {
            this.setProperty('scroll-behavior', scrollBehavior);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMargin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin');
        },
        /**
         *
         */
        set: function (scrollMargin) {
            this.setProperty('scroll-margin', scrollMargin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginBlock", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-block');
        },
        /**
         *
         */
        set: function (scrollMarginBlock) {
            this.setProperty('scroll-margin-block', scrollMarginBlock);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginBlockEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-block-end');
        },
        /**
         *
         */
        set: function (scrollMarginBlockEnd) {
            this.setProperty('scroll-margin-block-end', scrollMarginBlockEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginBlockStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-block-start');
        },
        /**
         *
         */
        set: function (scrollMarginBlockStart) {
            this.setProperty('scroll-margin-block-start', scrollMarginBlockStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginBottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-bottom');
        },
        /**
         *
         */
        set: function (scrollMarginBottom) {
            this.setProperty('scroll-margin-bottom', scrollMarginBottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginInline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-inline');
        },
        /**
         *
         */
        set: function (scrollMarginInline) {
            this.setProperty('scroll-margin-inline', scrollMarginInline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginInlineEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-inline-end');
        },
        /**
         *
         */
        set: function (scrollMarginInlineEnd) {
            this.setProperty('scroll-margin-inline-end', scrollMarginInlineEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginInlineStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-inline-start');
        },
        /**
         *
         */
        set: function (scrollMarginInlineStart) {
            this.setProperty('scroll-margin-inline-start', scrollMarginInlineStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginLeft", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-left');
        },
        /**
         *
         */
        set: function (scrollMarginLeft) {
            this.setProperty('scroll-margin-left', scrollMarginLeft);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginRight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-right');
        },
        /**
         *
         */
        set: function (scrollMarginRight) {
            this.setProperty('scroll-margin-right', scrollMarginRight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollMarginTop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-margin-top');
        },
        /**
         *
         */
        set: function (scrollMarginTop) {
            this.setProperty('scroll-margin-top', scrollMarginTop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPadding", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding');
        },
        /**
         *
         */
        set: function (scrollPadding) {
            this.setProperty('scroll-padding', scrollPadding);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingBlock", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-block');
        },
        /**
         *
         */
        set: function (scrollPaddingBlock) {
            this.setProperty('scroll-padding-block', scrollPaddingBlock);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingBlockEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-block-end');
        },
        /**
         *
         */
        set: function (scrollPaddingBlockEnd) {
            this.setProperty('scroll-padding-block-end', scrollPaddingBlockEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingBlockStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-block-start');
        },
        /**
         *
         */
        set: function (scrollPaddingBlockStart) {
            this.setProperty('scroll-padding-block-start', scrollPaddingBlockStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingBottom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-bottom');
        },
        /**
         *
         */
        set: function (scrollPaddingBottom) {
            this.setProperty('scroll-padding-bottom', scrollPaddingBottom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingInline", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-inline');
        },
        /**
         *
         */
        set: function (scrollPaddingInline) {
            this.setProperty('scroll-padding-inline', scrollPaddingInline);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingInlineEnd", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-inline-end');
        },
        /**
         *
         */
        set: function (scrollPaddingInlineEnd) {
            this.setProperty('scroll-padding-inline-end', scrollPaddingInlineEnd);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingInlineStart", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-inline-start');
        },
        /**
         *
         */
        set: function (scrollPaddingInlineStart) {
            this.setProperty('scroll-padding-inline-start', scrollPaddingInlineStart);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingLeft", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-left');
        },
        /**
         *
         */
        set: function (scrollPaddingLeft) {
            this.setProperty('scroll-padding-left', scrollPaddingLeft);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingRight", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-right');
        },
        /**
         *
         */
        set: function (scrollPaddingRight) {
            this.setProperty('scroll-padding-right', scrollPaddingRight);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollPaddingTop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-padding-top');
        },
        /**
         *
         */
        set: function (scrollPaddingTop) {
            this.setProperty('scroll-padding-top', scrollPaddingTop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollSnapAlign", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-snap-align');
        },
        /**
         *
         */
        set: function (scrollSnapAlign) {
            this.setProperty('scroll-snap-align', scrollSnapAlign);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollSnapStop", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-snap-stop');
        },
        /**
         *
         */
        set: function (scrollSnapStop) {
            this.setProperty('scroll-snap-stop', scrollSnapStop);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "scrollSnapType", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('scroll-snap-type');
        },
        /**
         *
         */
        set: function (scrollSnapType) {
            this.setProperty('scroll-snap-type', scrollSnapType);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "shapeImageThreshold", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('shape-image-threshold');
        },
        /**
         *
         */
        set: function (shapeImageThreshold) {
            this.setProperty('shape-image-threshold', shapeImageThreshold);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "shapeMargin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('shape-margin');
        },
        /**
         *
         */
        set: function (shapeMargin) {
            this.setProperty('shape-margin', shapeMargin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "shapeOutside", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('shape-outside');
        },
        /**
         *
         */
        set: function (shapeOutside) {
            this.setProperty('shape-outside', shapeOutside);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "shapeRendering", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('shape-rendering');
        },
        /**
         *
         */
        set: function (shapeRendering) {
            this.setProperty('shape-rendering', shapeRendering);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "size", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('size');
        },
        /**
         *
         */
        set: function (size) {
            this.setProperty('size', size);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "speak", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('speak');
        },
        /**
         *
         */
        set: function (speak) {
            this.setProperty('speak', speak);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "src", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('src');
        },
        /**
         *
         */
        set: function (src) {
            this.setProperty('src', src);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "stopColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stop-color');
        },
        /**
         *
         */
        set: function (stopColor) {
            this.setProperty('stop-color', stopColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "stopOpacity", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stop-opacity');
        },
        /**
         *
         */
        set: function (stopOpacity) {
            this.setProperty('stop-opacity', stopOpacity);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "stroke", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke');
        },
        /**
         *
         */
        set: function (stroke) {
            this.setProperty('stroke', stroke);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeDasharray", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-dasharray');
        },
        /**
         *
         */
        set: function (strokeDasharray) {
            this.setProperty('stroke-dasharray', strokeDasharray);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeDashoffset", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-dashoffset');
        },
        /**
         *
         */
        set: function (strokeDashoffset) {
            this.setProperty('stroke-dashoffset', strokeDashoffset);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeLinecap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-linecap');
        },
        /**
         *
         */
        set: function (strokeLinecap) {
            this.setProperty('stroke-linecap', strokeLinecap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeLinejoin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-linejoin');
        },
        /**
         *
         */
        set: function (strokeLinejoin) {
            this.setProperty('stroke-linejoin', strokeLinejoin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeMiterlimit", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-miterlimit');
        },
        /**
         *
         */
        set: function (strokeMiterlimit) {
            this.setProperty('stroke-miterlimit', strokeMiterlimit);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeOpacity", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-opacity');
        },
        /**
         *
         */
        set: function (strokeOpacity) {
            this.setProperty('stroke-opacity', strokeOpacity);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "strokeWidth", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('stroke-width');
        },
        /**
         *
         */
        set: function (strokeWidth) {
            this.setProperty('stroke-width', strokeWidth);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "syntax", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('syntax');
        },
        /**
         *
         */
        set: function (syntax) {
            this.setProperty('syntax', syntax);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "tabSize", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('tab-size');
        },
        /**
         *
         */
        set: function (tabSize) {
            this.setProperty('tab-size', tabSize);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "tableLayout", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('table-layout');
        },
        /**
         *
         */
        set: function (tableLayout) {
            this.setProperty('table-layout', tableLayout);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textAlign", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-align');
        },
        /**
         *
         */
        set: function (textAlign) {
            this.setProperty('text-align', textAlign);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textAlignLast", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-align-last');
        },
        /**
         *
         */
        set: function (textAlignLast) {
            this.setProperty('text-align-last', textAlignLast);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textAnchor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-anchor');
        },
        /**
         *
         */
        set: function (textAnchor) {
            this.setProperty('text-anchor', textAnchor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textCombineUpright", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-combine-upright');
        },
        /**
         *
         */
        set: function (textCombineUpright) {
            this.setProperty('text-combine-upright', textCombineUpright);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textDecoration", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-decoration');
        },
        /**
         *
         */
        set: function (textDecoration) {
            this.setProperty('text-decoration', textDecoration);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textDecorationColor", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-decoration-color');
        },
        /**
         *
         */
        set: function (textDecorationColor) {
            this.setProperty('text-decoration-color', textDecorationColor);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textDecorationLine", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-decoration-line');
        },
        /**
         *
         */
        set: function (textDecorationLine) {
            this.setProperty('text-decoration-line', textDecorationLine);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textDecorationSkipInk", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-decoration-skip-ink');
        },
        /**
         *
         */
        set: function (textDecorationSkipInk) {
            this.setProperty('text-decoration-skip-ink', textDecorationSkipInk);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textDecorationStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-decoration-style');
        },
        /**
         *
         */
        set: function (textDecorationStyle) {
            this.setProperty('text-decoration-style', textDecorationStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textIndent", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-indent');
        },
        /**
         *
         */
        set: function (textIndent) {
            this.setProperty('text-indent', textIndent);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textOrientation", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-orientation');
        },
        /**
         *
         */
        set: function (textOrientation) {
            this.setProperty('text-orientation', textOrientation);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textOverflow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-overflow');
        },
        /**
         *
         */
        set: function (textOverflow) {
            this.setProperty('text-overflow', textOverflow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textRendering", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-rendering');
        },
        /**
         *
         */
        set: function (textRendering) {
            this.setProperty('text-rendering', textRendering);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textShadow", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-shadow');
        },
        /**
         *
         */
        set: function (textShadow) {
            this.setProperty('text-shadow', textShadow);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textSizeAdjust", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-size-adjust');
        },
        /**
         *
         */
        set: function (textSizeAdjust) {
            this.setProperty('text-size-adjust', textSizeAdjust);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textTransform", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-transform');
        },
        /**
         *
         */
        set: function (textTransform) {
            this.setProperty('text-transform', textTransform);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "textUnderlinePosition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('text-underline-position');
        },
        /**
         *
         */
        set: function (textUnderlinePosition) {
            this.setProperty('text-underline-position', textUnderlinePosition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "top", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('top');
        },
        /**
         *
         */
        set: function (top) {
            this.setProperty('top', top);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "touchAction", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('touch-action');
        },
        /**
         *
         */
        set: function (touchAction) {
            this.setProperty('touch-action', touchAction);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transform", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transform');
        },
        /**
         *
         */
        set: function (transform) {
            this.setProperty('transform', transform);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transformBox", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transform-box');
        },
        /**
         *
         */
        set: function (transformBox) {
            this.setProperty('transform-box', transformBox);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transformOrigin", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transform-origin');
        },
        /**
         *
         */
        set: function (transformOrigin) {
            this.setProperty('transform-origin', transformOrigin);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transformStyle", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transform-style');
        },
        /**
         *
         */
        set: function (transformStyle) {
            this.setProperty('transform-style', transformStyle);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transition", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transition');
        },
        /**
         *
         */
        set: function (transition) {
            this.setProperty('transition', transition);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transitionDelay", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transition-delay');
        },
        /**
         *
         */
        set: function (transitionDelay) {
            this.setProperty('transition-delay', transitionDelay);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transitionDuration", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transition-duration');
        },
        /**
         *
         */
        set: function (transitionDuration) {
            this.setProperty('transition-duration', transitionDuration);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transitionProperty", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transition-property');
        },
        /**
         *
         */
        set: function (transitionProperty) {
            this.setProperty('transition-property', transitionProperty);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "transitionTimingFunction", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('transition-timing-function');
        },
        /**
         *
         */
        set: function (transitionTimingFunction) {
            this.setProperty('transition-timing-function', transitionTimingFunction);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "unicodeBidi", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('unicode-bidi');
        },
        /**
         *
         */
        set: function (unicodeBidi) {
            this.setProperty('unicode-bidi', unicodeBidi);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "unicodeRange", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('unicode-range');
        },
        /**
         *
         */
        set: function (unicodeRange) {
            this.setProperty('unicode-range', unicodeRange);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "userSelect", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('user-select');
        },
        /**
         *
         */
        set: function (userSelect) {
            this.setProperty('user-select', userSelect);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "userZoom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('user-zoom');
        },
        /**
         *
         */
        set: function (userZoom) {
            this.setProperty('user-zoom', userZoom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "vectorEffect", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('vector-effect');
        },
        /**
         *
         */
        set: function (vectorEffect) {
            this.setProperty('vector-effect', vectorEffect);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "verticalAlign", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('vertical-align');
        },
        /**
         *
         */
        set: function (verticalAlign) {
            this.setProperty('vertical-align', verticalAlign);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "visibility", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('visibility');
        },
        /**
         *
         */
        set: function (visibility) {
            this.setProperty('visibility', visibility);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "whiteSpace", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('white-space');
        },
        /**
         *
         */
        set: function (whiteSpace) {
            this.setProperty('white-space', whiteSpace);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "widows", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('widows');
        },
        /**
         *
         */
        set: function (widows) {
            this.setProperty('widows', widows);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "width", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('width');
        },
        /**
         *
         */
        set: function (width) {
            this.setProperty('width', width);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "willChange", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('will-change');
        },
        /**
         *
         */
        set: function (willChange) {
            this.setProperty('will-change', willChange);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "wordBreak", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('word-break');
        },
        /**
         *
         */
        set: function (wordBreak) {
            this.setProperty('word-break', wordBreak);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "wordSpacing", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('word-spacing');
        },
        /**
         *
         */
        set: function (wordSpacing) {
            this.setProperty('word-spacing', wordSpacing);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "wordWrap", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('word-wrap');
        },
        /**
         *
         */
        set: function (wordWrap) {
            this.setProperty('word-wrap', wordWrap);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "writingMode", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('writing-mode');
        },
        /**
         *
         */
        set: function (writingMode) {
            this.setProperty('writing-mode', writingMode);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "x", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('x');
        },
        /**
         *
         */
        set: function (x) {
            this.setProperty('x', x);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "y", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('y');
        },
        /**
         *
         */
        set: function (y) {
            this.setProperty('y', y);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "zIndex", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('z-index');
        },
        /**
         *
         */
        set: function (zIndex) {
            this.setProperty('z-index', zIndex);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "zoom", {
        /**
         *
         */
        get: function () {
            return this.getPropertyValue('zoom');
        },
        /**
         *
         */
        set: function (zoom) {
            this.setProperty('zoom', zoom);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CSSStyleDeclaration.prototype, "cssText", {
        /* eslint-enable require-jsdoc */
        /**
         * Returns the style decleration as a CSS text.
         *
         * @returns CSS text.
         */
        get: function () {
            var style = this._attributes['style'];
            if (style && style.value) {
                return style.value;
            }
            return '';
        },
        /**
         * Sets CSS text.
         *
         * @param cssText CSS text.
         */
        set: function (cssText) {
            if (cssText) {
                if (!this._attributes['style']) {
                    this._attributes['style'] = new Attr_1.default();
                    this._attributes['style'].name = 'style';
                }
                var parts = cssText.split(';');
                var newStyle = [];
                var index = 0;
                for (var i = 0; i < this.length; i++) {
                    delete this[i];
                }
                for (var _i = 0, parts_2 = parts; _i < parts_2.length; _i++) {
                    var part = parts_2[_i];
                    if (part) {
                        var _a = part.trim().split(':'), name = _a[0], value = _a[1];
                        newStyle.push(name + ": " + value.trim() + ";");
                        this[index] = name;
                        index++;
                    }
                }
                this.length = index;
                this._attributes['style'].value = newStyle.join(' ');
            }
            else {
                delete this._attributes['style'];
                for (var i = 0; i < this.length; i++) {
                    delete this[i];
                }
                this.length = 0;
            }
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns item.
     *
     * @param index Index.
     * @returns Item.
     */
    CSSStyleDeclaration.prototype.item = function (index) {
        return this[String(index)] || '';
    };
    /**
     * Set a property.
     *
     * @param propertyName Property name.
     * @param value Value. Must not contain "!important" as that should be set using the priority parameter.
     * @param [priority] Can be "important", or an empty string.
     */
    CSSStyleDeclaration.prototype.setProperty = function (propertyName, value, priority) {
        if (priority === void 0) { priority = ''; }
        if (!value) {
            this.removeProperty(propertyName);
            return;
        }
        if (!this._attributes['style']) {
            this._attributes['style'] = new Attr_1.default();
            this._attributes['style'].name = 'style';
        }
        var style = this._attributes['style'];
        var newStyle = [];
        var index = 0;
        var isExisting = false;
        if (style && style.value) {
            var parts = style.value.split(';');
            for (var _i = 0, parts_3 = parts; _i < parts_3.length; _i++) {
                var part = parts_3[_i];
                if (part) {
                    var _a = part.trim().split(':'), name = _a[0], existingValue = _a[1];
                    if (name === propertyName) {
                        newStyle.push(name + ": " + value + ";");
                        isExisting = true;
                    }
                    else {
                        newStyle.push(name + ": " + existingValue.trim() + ";");
                    }
                    this[index] = name;
                    index++;
                }
            }
        }
        if (!isExisting) {
            newStyle.push(propertyName + ": " + value + (priority ? '' + priority : '') + ";");
            this[index] = propertyName;
            index++;
        }
        this._attributes['style'].value = newStyle.join(' ');
        this.length = index;
    };
    /**
     * Removes a property.
     *
     * @param propertyName Property name in kebab case.
     * @param value Value. Must not contain "!important" as that should be set using the priority parameter.
     * @param [priority] Can be "important", or an empty string.
     */
    CSSStyleDeclaration.prototype.removeProperty = function (propertyName) {
        var style = this._attributes['style'];
        var newStyle = [];
        var hasProperty = false;
        var index = 0;
        if (style && style.value) {
            var parts = style.value.split(';');
            for (var _i = 0, parts_4 = parts; _i < parts_4.length; _i++) {
                var part = parts_4[_i];
                if (part) {
                    var _a = part.trim().split(':'), name = _a[0], value = _a[1];
                    if (name !== propertyName) {
                        newStyle.push(name + ": " + value.trim() + ";");
                        this[index] = name;
                        index++;
                        hasProperty = true;
                    }
                }
            }
        }
        if (newStyle.length) {
            this._attributes['style'].value = newStyle.join(' ');
        }
        else {
            delete this._attributes['style'];
        }
        if (hasProperty) {
            delete this[index];
        }
        this.length = index;
    };
    /**
     * Returns a property.
     *
     * @param propertyName Property name in kebab case.
     * @returns Property value.
     */
    CSSStyleDeclaration.prototype.getPropertyValue = function (propertyName) {
        if (this._computedStyleElement && !this._computedStyleElement.isConnected) {
            return '';
        }
        var style = this._attributes['style'];
        if (style && style.value) {
            var parts = style.value.split(';');
            for (var _i = 0, parts_5 = parts; _i < parts_5.length; _i++) {
                var part = parts_5[_i];
                if (part) {
                    var _a = part.trim().split(':'), name = _a[0], value = _a[1];
                    if (name === propertyName) {
                        return value.trim();
                    }
                }
            }
        }
        return '';
    };
    return CSSStyleDeclaration;
}());
exports.default = CSSStyleDeclaration;
