// happy-dom/lib/config/ElementClass.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var HTMLTemplateElement_1 = __importDefault(require("../nodes/html-template-element/HTMLTemplateElement"));
var HTMLFormElement_1 = __importDefault(require("../nodes/html-form-element/HTMLFormElement"));
var HTMLElement_1 = __importDefault(require("../nodes/html-element/HTMLElement"));
var HTMLInputElement_1 = __importDefault(require("../nodes/html-input-element/HTMLInputElement"));
var HTMLTextAreaElement_1 = __importDefault(require("../nodes/html-text-area-element/HTMLTextAreaElement"));
var SVGSVGElement_1 = __importDefault(require("../nodes/svg-element/SVGSVGElement"));
var SVGElement_1 = __importDefault(require("../nodes/svg-element/SVGElement"));
var HTMLScriptElement_1 = __importDefault(require("../nodes/html-script-element/HTMLScriptElement"));
var HTMLImageElement_1 = __importDefault(require("../nodes/html-image-element/HTMLImageElement"));
var HTMLStyleElement_1 = __importDefault(require("../nodes/html-style-element/HTMLStyleElement"));
var HTMLLinkElement_1 = __importDefault(require("../nodes/html-link-element/HTMLLinkElement"));
exports.default = {
    HTMLTemplateElement: HTMLTemplateElement_1.default,
    HTMLFormElement: HTMLFormElement_1.default,
    HTMLInputElement: HTMLInputElement_1.default,
    HTMLTextAreaElement: HTMLTextAreaElement_1.default,
    HTMLElement: HTMLElement_1.default,
    HTMLHeadElement: HTMLElement_1.default,
    HTMLTitleElement: HTMLElement_1.default,
    HTMLBaseElement: HTMLElement_1.default,
    HTMLLinkElement: HTMLLinkElement_1.default,
    HTMLMetaElement: HTMLElement_1.default,
    HTMLStyleElement: HTMLStyleElement_1.default,
    HTMLBodyElement: HTMLElement_1.default,
    HTMLHeadingElement: HTMLElement_1.default,
    HTMLParagraphElement: HTMLElement_1.default,
    HTMLHRElement: HTMLElement_1.default,
    HTMLPreElement: HTMLElement_1.default,
    HTMLUListElement: HTMLElement_1.default,
    HTMLOListElement: HTMLElement_1.default,
    HTMLLIElement: HTMLElement_1.default,
    HTMLMenuElement: HTMLElement_1.default,
    HTMLDListElement: HTMLElement_1.default,
    HTMLDivElement: HTMLElement_1.default,
    HTMLAnchorElement: HTMLElement_1.default,
    HTMLAreaElement: HTMLElement_1.default,
    HTMLBRElement: HTMLElement_1.default,
    HTMLButtonElement: HTMLElement_1.default,
    HTMLCanvasElement: HTMLElement_1.default,
    HTMLDataElement: HTMLElement_1.default,
    HTMLDataListElement: HTMLElement_1.default,
    HTMLDetailsElement: HTMLElement_1.default,
    HTMLDialogElement: HTMLElement_1.default,
    HTMLDirectoryElement: HTMLElement_1.default,
    HTMLFieldSetElement: HTMLElement_1.default,
    HTMLFontElement: HTMLElement_1.default,
    HTMLHtmlElement: HTMLElement_1.default,
    HTMLImageElement: HTMLImageElement_1.default,
    HTMLLabelElement: HTMLElement_1.default,
    HTMLLegendElement: HTMLElement_1.default,
    HTMLMapElement: HTMLElement_1.default,
    HTMLMarqueeElement: HTMLElement_1.default,
    HTMLMediaElement: HTMLElement_1.default,
    HTMLMeterElement: HTMLElement_1.default,
    HTMLModElement: HTMLElement_1.default,
    HTMLOptGroupElement: HTMLElement_1.default,
    HTMLOptionElement: HTMLElement_1.default,
    HTMLOutputElement: HTMLElement_1.default,
    HTMLPictureElement: HTMLElement_1.default,
    HTMLProgressElement: HTMLElement_1.default,
    HTMLQuoteElement: HTMLElement_1.default,
    HTMLScriptElement: HTMLScriptElement_1.default,
    HTMLSelectElement: HTMLElement_1.default,
    HTMLSlotElement: HTMLElement_1.default,
    HTMLSourceElement: HTMLElement_1.default,
    HTMLSpanElement: HTMLElement_1.default,
    HTMLTableCaptionElement: HTMLElement_1.default,
    HTMLTableCellElement: HTMLElement_1.default,
    HTMLTableColElement: HTMLElement_1.default,
    HTMLTableElement: HTMLElement_1.default,
    HTMLTimeElement: HTMLElement_1.default,
    HTMLTableRowElement: HTMLElement_1.default,
    HTMLTableSectionElement: HTMLElement_1.default,
    HTMLUnknownElement: HTMLElement_1.default,
    HTMLFrameElement: HTMLElement_1.default,
    HTMLFrameSetElement: HTMLElement_1.default,
    HTMLIFrameElement: HTMLElement_1.default,
    HTMLEmbedElement: HTMLElement_1.default,
    HTMLObjectElement: HTMLElement_1.default,
    HTMLParamElement: HTMLElement_1.default,
    HTMLVideoElement: HTMLElement_1.default,
    HTMLAudioElement: HTMLElement_1.default,
    HTMLTrackElement: HTMLElement_1.default,
    SVGSVGElement: SVGSVGElement_1.default,
    SVGElement: SVGElement_1.default
};
