// happy-dom/lib/nodes/document-fragment/DocumentFragment.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Node_1 = __importDefault(require("../node/Node"));
var QuerySelector_1 = __importDefault(require("../../query-selector/QuerySelector"));
var ParentNodeUtility_1 = __importDefault(require("../parent-node/ParentNodeUtility"));
var HTMLCollectionFactory_1 = __importDefault(require("../element/HTMLCollectionFactory"));
/**
 * DocumentFragment.
 */
var DocumentFragment = /** @class */ (function (_super) {
    __extends(DocumentFragment, _super);
    function DocumentFragment() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.nodeType = Node_1.default.DOCUMENT_FRAGMENT_NODE;
        _this.children = HTMLCollectionFactory_1.default.create();
        return _this;
    }
    Object.defineProperty(DocumentFragment.prototype, "childElementCount", {
        /**
         * Last element child.
         *
         * @returns Element.
         */
        get: function () {
            return this.children.length;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DocumentFragment.prototype, "firstElementChild", {
        /**
         * First element child.
         *
         * @returns Element.
         */
        get: function () {
            return this.children ? this.children[0] || null : null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DocumentFragment.prototype, "lastElementChild", {
        /**
         * Last element child.
         *
         * @returns Element.
         */
        get: function () {
            return this.children ? this.children[this.children.length - 1] || null : null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DocumentFragment.prototype, "textContent", {
        /**
         * Get text value of children.
         *
         * @returns Text content.
         */
        get: function () {
            var result = '';
            for (var _i = 0, _a = this.childNodes; _i < _a.length; _i++) {
                var childNode = _a[_i];
                if (childNode.nodeType === Node_1.default.ELEMENT_NODE || childNode.nodeType === Node_1.default.TEXT_NODE) {
                    result += childNode.textContent;
                }
            }
            return result;
        },
        /**
         * Sets text content.
         *
         * @param textContent Text content.
         */
        set: function (textContent) {
            for (var _i = 0, _a = this.childNodes.slice(); _i < _a.length; _i++) {
                var child = _a[_i];
                this.removeChild(child);
            }
            this.appendChild(this.ownerDocument.createTextNode(textContent));
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Inserts a set of Node objects or DOMString objects after the last child of the ParentNode. DOMString objects are inserted as equivalent Text nodes.
     *
     * @param nodes List of Node or DOMString.
     */
    DocumentFragment.prototype.append = function () {
        var nodes = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            nodes[_i] = arguments[_i];
        }
        ParentNodeUtility_1.default.append.apply(ParentNodeUtility_1.default, __spreadArray([this], nodes));
    };
    /**
     * Inserts a set of Node objects or DOMString objects before the first child of the ParentNode. DOMString objects are inserted as equivalent Text nodes.
     *
     * @param nodes List of Node or DOMString.
     */
    DocumentFragment.prototype.prepend = function () {
        var nodes = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            nodes[_i] = arguments[_i];
        }
        ParentNodeUtility_1.default.prepend.apply(ParentNodeUtility_1.default, __spreadArray([this], nodes));
    };
    /**
     * Replaces the existing children of a node with a specified new set of children.
     *
     * @param nodes List of Node or DOMString.
     */
    DocumentFragment.prototype.replaceChildren = function () {
        var nodes = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            nodes[_i] = arguments[_i];
        }
        ParentNodeUtility_1.default.replaceChildren.apply(ParentNodeUtility_1.default, __spreadArray([this], nodes));
    };
    /**
     * Query CSS selector to find matching nodes.
     *
     * @param selector CSS selector.
     * @returns Matching elements.
     */
    DocumentFragment.prototype.querySelectorAll = function (selector) {
        return QuerySelector_1.default.querySelectorAll(this, selector);
    };
    /**
     * Query CSS Selector to find matching node.
     *
     * @param selector CSS selector.
     * @returns Matching element.
     */
    DocumentFragment.prototype.querySelector = function (selector) {
        return QuerySelector_1.default.querySelector(this, selector);
    };
    /**
     * Returns an element by ID.
     *
     * @param id ID.
     * @returns Matching element.
     */
    DocumentFragment.prototype.getElementById = function (id) {
        return ParentNodeUtility_1.default.getElementById(this, id);
    };
    /**
     * Clones a node.
     *
     * @override
     * @param [deep=false] "true" to clone deep.
     * @returns Cloned node.
     */
    DocumentFragment.prototype.cloneNode = function (deep) {
        if (deep === void 0) { deep = false; }
        var clone = _super.prototype.cloneNode.call(this, deep);
        if (deep) {
            for (var _i = 0, _a = clone.childNodes; _i < _a.length; _i++) {
                var node = _a[_i];
                if (node.nodeType === Node_1.default.ELEMENT_NODE) {
                    clone.children.push(node);
                }
            }
        }
        return clone;
    };
    /**
     * Append a child node to childNodes.
     *
     * @override
     * @param  node Node to append.
     * @returns Appended node.
     */
    DocumentFragment.prototype.appendChild = function (node) {
        // If the type is DocumentFragment, then the child nodes of if it should be moved instead of the actual node.
        // See: https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment
        if (node.nodeType !== Node_1.default.DOCUMENT_FRAGMENT_NODE) {
            if (node.parentNode && node.parentNode['children']) {
                var index = node.parentNode['children'].indexOf(node);
                if (index !== -1) {
                    node.parentNode['children'].splice(index, 1);
                }
            }
            if (node !== this && node.nodeType === Node_1.default.ELEMENT_NODE) {
                this.children.push(node);
            }
        }
        return _super.prototype.appendChild.call(this, node);
    };
    /**
     * Remove Child element from childNodes array.
     *
     * @override
     * @param node Node to remove.
     */
    DocumentFragment.prototype.removeChild = function (node) {
        if (node.nodeType === Node_1.default.ELEMENT_NODE) {
            var index = this.children.indexOf(node);
            if (index !== -1) {
                this.children.splice(index, 1);
            }
        }
        return _super.prototype.removeChild.call(this, node);
    };
    /**
     * Inserts a node before another.
     *
     * @override
     * @param newNode Node to insert.
     * @param [referenceNode] Node to insert before.
     * @returns Inserted node.
     */
    DocumentFragment.prototype.insertBefore = function (newNode, referenceNode) {
        var returnValue = _super.prototype.insertBefore.call(this, newNode, referenceNode);
        // If the type is DocumentFragment, then the child nodes of if it should be moved instead of the actual node.
        // See: https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment
        if (newNode.nodeType !== Node_1.default.DOCUMENT_FRAGMENT_NODE) {
            if (newNode.parentNode && newNode.parentNode['children']) {
                var index = newNode.parentNode['children'].indexOf(newNode);
                if (index !== -1) {
                    newNode.parentNode['children'].splice(index, 1);
                }
            }
            this.children.length = 0;
            for (var _i = 0, _a = this.childNodes; _i < _a.length; _i++) {
                var node = _a[_i];
                if (node.nodeType === Node_1.default.ELEMENT_NODE) {
                    this.children.push(node);
                }
            }
        }
        return returnValue;
    };
    return DocumentFragment;
}(Node_1.default));
exports.default = DocumentFragment;
